[#168] Make activities to be matched by autocomplete configurable

elbenfreund opened the pull request:
This PR introduces a new config key ``autocomplete_activities_offset`` which is expected to be an integer.
When autocompleting raw facts a set of 'recent activities' is fetched. This new config value will determine the number of days before today that should be the start for this collection of 'recent activities'.
We also add another new config key ``autocomplete_split_activity`` which is a boolean and governs if a typed string will be matched against ``activity.name`` and ``activity.category.name`` individually or in conjunction.

Closes: #168

elbenfreund committed:
Make HamsterGTK._config a public attribute

This commit just provides an 'alias' for the still used private
attribute. It is up to issue #169 to replace all legacy references to
the former private ``_config`` attribute with ones to the new public
``config``.

elbenfreund committed:
Pass app config to RawFactCompletion

In order to make the apps config available for the ``RawFactCompletion``
in our ``StartTracking`` screen we have to pass it thorugh all relevant
parrents. We do so explicitly and do not simply pass the app as a whole
in oder to be more transparent about what we do and do not need.
This also means adjusting some tests/fixtures.

elbenfreund committed:
Add 'autocomplete_activities_offset' config value

elbenfreund committed:
Use 'autocomplete_activities_offset' for autocomplete

Ref: #168

elbenfreund committed:
Integrate 'autocomplete_activities_offset' in Preferences

Ref: #168

jtojnar commented:
How will this work in the context of hamster-lib’s config manager (i.e. [LIB-235](https://projecthamster.atlassian.net/browse/LIB-235))?

jtojnar commented:
Actually, I think the config helpers are still something that every client should implement by itself. For example hamster-gtk should go for #21.

jtojnar commented on hamster_gtk/hamster_gtk.py:
```diff
@@ -275,6 +276,8 @@ def _get_default_config(self):
             'db_engine': 'sqlite',
             'db_path': os.path.join(appdirs.user_data_dir, 'hamster-gtk.sqlite'),
+            # Frontend
+            'autocomplete_activities_offset': 30,
```
I do not really like this name. Something like `activity_autocomplete_range` sounds slightly better but still not optimal.

elbenfreund replied:
I do not like the name either, but fail to come up with somehting better.
I think ``activity_autocomplete_range`` is worthe than starting with "autocomplete" as the latter limits the "namespace" more than "activity" and hence already makes the name more specific towards its use. Also it is not used  to "autocomplete activities" but to "fetch activities for autocompletion in general". I agree though that ``..._range`` or maybe even better ``..._days`` would be clearer.

jtojnar replied:
Either way, I would use [singular](https://en.wikipedia.org/wiki/English_plurals#Nouns_used_attributively).

elbenfreund commented:
Whilst I agree this is something we need to talk about I would like to avoid getting distracted from the PR at hand.
As far as this PR is concerned ``hamster-gtk`` ist config file based and this PR provides a way to deal with this in a consistent way. If there is to emerge a change / extention in the way configuration is done this will be a different matter.

Do you have any other feedback/comments/improvements about this PR or does this mean you are ok with the actual code?

ps: with regards to LIB-235: as far as this PR is concerned this issue is still unresolved. However, even if ``hamster-gtk`` would use the infrastructure provided by ``hamster-lib`` this particular config setting is a "frontend" setting and as such totally unrelated to whatever ``hamster-lib`` would aim to provide...

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -265,8 +268,9 @@ def _get_activities(self):
         for autocomplete suggestions.
         """
         today = datetime.date.today()
+        start = today - datetime.timedelta(self._config['autocomplete_activities_offset'])
```
It might be better to explicitly specify that it is the days argument.

jtojnar commented on hamster_gtk/hamster_gtk.py:
```diff
@@ -363,11 +373,15 @@ def get_db_config():
             return result
 
+        def get_autocomplete_activities_offset():
+            return cp_instance.getint('Frontend', 'autocomplete_activities_offset')
```
When existing config is used, selfoss crashes on`backports.configparser.NoSectionError: No section: 'Frontend'`

elbenfreund replied:
Yes. This is somewhat expected / desired. Right now all config keys are mandatory. so a preexisting config without all relevant keys is invalid. As we are not API stable at this point its acceptable that we crash with "invalid" older configs.

Before a major release we probably should provide default fallbacks for non supplied config keys though.

jtojnar commented on hamster_gtk/misc/widgets/raw_fact_entry.py:
```diff
@@ -60,22 +60,24 @@ def _get_segment_boundaries(segment, match):
     """A custom entry widgets that provides ``raw fact`` specific autocompletion behaviour."""
 
-    def __init__(self, controller, split_activity_autocomplete=False, *args, **kwargs):
+    def __init__(self, controller, config, split_activity_autocomplete=False, *args, **kwargs):
```
Changes in the preferences do not propagate – restart is needed.

jtojnar commented on hamster_gtk/preferences/preferences_dialog.py:
```diff
@@ -73,6 +73,10 @@ def __init__(self, parent, app, initial, *args, **kwargs):
                 ('db_path', (_('DB _Path'), ComboFileChooser())),
                 ('tmpfile_path', (_('_Temporary file'), ComboFileChooser())),
             ]))),
+            (_('Misc'), LabelledWidgetsGrid(collections.OrderedDict([
```
I would rather see Miscellaneous.

elbenfreund commented on hamster_gtk/hamster_gtk.py:
```diff
@@ -363,11 +373,15 @@ def get_db_config():
             return result
 
+        def get_autocomplete_activities_offset():
+            return cp_instance.getint('Frontend', 'autocomplete_activities_offset')
```
Yes. This is somewhat expected / desired. Right now all config keys are mandatory. so a preexisting config without all relevant keys is invalid. As we are not API stable at this point its acceptable that we crash with "invalid" older configs.

Before a major release we probably should provide default fallbacks for non supplied config keys though.

elbenfreund committed:
Rename 'Misc' preferencecs to 'Miscellaneous'

Ref: #168

elbenfreund committed:
Use explicit kwarg for 'timedelta' call

Ref: #168

elbenfreund committed:
Pass 'app' to tracking screen and its children

In order to have access to potentially updated app configs (esp. for the
rawfactentries) we return to pass the app itself to the screen and its
major widgets instead of just the current controler and config instance.

Ref: #168

elbenfreund committed:
Refactor and rename '_populate_stores'

elbenfreund committed:
Completion store population clears existing data

elbenfreund commented:
Ready again

elbenfreund committed:
CurrentFactBox only requires controller

We do not need the app instance for the current fact box, the controller
suffices.

jtojnar commented:
When trying to save a fact, I get “'HamsterControl' object has no attribute 'controller'”.

elbenfreund commented:
yes. sorry about that. should be fixed

elbenfreund committed:
Fix screen tests.

After changing from passing the app to just the controller for
CurrentFactBox its test were not adjusted and broke This commit fixes
that

elbenfreund committed:
Add 'autocomplete_split_activity' config option

This adds a new option to the config that will be used to covern
autocomplete behaviour. In order to represent this boolean value in the
preference screen a custom ``HamsterCheckButton`` has been added.

elbenfreund committed:
Use 'autocomplete_split_activity'

jtojnar commented on hamster_gtk/preferences/preferences_dialog.py:
```diff
@@ -76,6 +78,8 @@ def __init__(self, parent, app, initial, *args, **kwargs):
             (_('Miscellaneous'), LabelledWidgetsGrid(collections.OrderedDict([
                 ('autocomplete_activities_offset', (_("Autocomplete Activities Offset"),
                     HamsterSpinButton(SimpleAdjustment(0, GObject.G_MAXDOUBLE, 1)))),
+                ('autocomplete_split_activity', (_("Autocomplete Split Activity"),
```
“Autocomplete activities and categories separately” would be probably more meaningful.

jtojnar commented on hamster_gtk/preferences/preferences_dialog.py:
```diff
@@ -32,7 +32,9 @@
                                              HamsterComboBoxText,
                                              HamsterSpinButton,
-                                             SimpleAdjustment, TimeEntry)
+                                             HamsterCheckButton,
```
Isort

jtojnar commented on hamster_gtk/preferences/widgets/__init__.py:
```diff
@@ -21,4 +21,5 @@
 from .config_widget import ConfigWidget  # NOQA
 from .hamster_combo_box_text import HamsterComboBoxText  # NOQA
 from .hamster_spin_button import HamsterSpinButton, SimpleAdjustment  # NOQA
+from .hamster_check_button import HamsterCheckButton  # NOQA
```
Isort

jtojnar commented on hamster_gtk/preferences/widgets/hamster_check_button.py:
```diff
@@ -0,0 +1,58 @@
+# You should have received a copy of the GNU General Public License
+# along with 'hamster-gtk'.  If not, see <http://www.gnu.org/licenses/>.
+
+"""This module provides ComboBoxText widget extended to store preferences."""
```
wrong description

Commit:
Fix README typo

Closes: #129

elbenfreund committed:
Improving import order

elbenfreund committed:
Refactor phrasing in preference dialog

elbenfreund committed:
Correct docstring

elbenfreund commented:
Ready

elbenfreund committed:
Fix line length and missing blank line.

elbenfreund committed:
Use Switch instead of CheckButton in preferencecs

elbenfreund committed:
Rename 'autocomplete_activities_offset' to 'range'

jtojnar commented on hamster_gtk/preferences/preferences_dialog.py:
```diff
@@ -75,7 +75,7 @@ def __init__(self, parent, app, initial, *args, **kwargs):
             ]))),
             (_('Miscellaneous'), LabelledWidgetsGrid(collections.OrderedDict([
-                ('autocomplete_activities_offset', (_("Autocomplete Activities Offset"),
+                ('autocomplete_activities_range', (_("Autocomplete Activities Range"),
```
Something like “Autocomplete activities newer than” would be more descriptive.

elbenfreund replied:
While I agree that it may be slightly better it is still not really great. At some point I think some additional (optional/hover...) description would be useful for most settings I guess. I suggest we revisit the naming choices then instead. Also, this may have the benefit of ``hamster-gtk`` having a bigger audience that may help coming up with better descriptions.

elbenfreund commented:
ready
